### PR TITLE
docs: improve documentation

### PR DESCRIPTION
**Pull request by fjl**
This is basically a rewrite of the hive documentation. I've split up the docs into multiple files.
README was getting a bit long and contained a mix of tutorials for different things and the
API reference. In the new version, each major concept has its own page.

To review this, it's probably best to view the branch in the main GitHub UI:

   https://github.com/ethereum/hive/tree/docs-improvement

### CODE REVIEW

File: docs/clients.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+When the simulation requests a client instance, hive creates a docker container from the
+client image. The simulator can customize the container by passing environment variables
+with prefix `HIVE_`. It may also upload files into the container before it starts. Once
+the container is created, hive simply runs the entry point defined in the `Dockerfile`.
+
+For all client containers, hive waits for TCP port 8545 to open before considering the
+client ready for use by the simulator. If the client container does not open this port
+within a certain timeout, hive assumes the client has failed to start.
+
+Environment variables and files interpreted by thef entry point define a 'protocol'
```
Comment (renaynay) [inline docs/clients.md:37]:
```suggestion
Environment variables and files interpreted by the entry point define a 'protocol'
```

File: docs/clients.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+between the simulator and client. While hive itself does not require support for any
+specific variables or files, simulators usually expect client containers to be
+configurable in certain ways. In order to run tests against multiple Ethereum clients, for
+example, the simulator needs to be able to configure all clients for a specific blockchain
+and make them join the peer-to-peer network used for testing.
+
+## Eth1 Client Requirements
+
+This section describes the requirements for Ethereum 1.x client wrappers in hive. Client
+entry point scripts must support this interface in order to be tested by existing Ethereum
```
Comment (renaynay) [inline docs/clients.md:47]:
```suggestion
entrypoint scripts must support this interface in order to be tested by existing Ethereum
```

File: docs/clients.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+1.x-specific simulators.
+
+Clients must provide JSON-RPC over HTTP on TCP port 8545. They may also support JSON-RPC
+over WebSocket on port 8546, but this is not strictly required.
+
+### Files
+
+The simulator may customize client start by placing these files into the client container:
```
Comment (renaynay) [inline docs/clients.md:55]:
```suggestion
The simulator may customize client start-up by placing these files into the client container:
```

File: docs/clients.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+- `/genesis.json` contains Ethereum genesis state in the JSON format used by Geth. This
+  file is mandatory.
+- `/chain.rlp` contains RLP-encoded blocks to import before startup.
+- `/blocks/` directory containg `.rlp` files.
```
Comment (renaynay) [inline docs/clients.md:60]:
```suggestion
- `/blocks/` directory containing `.rlp` files.
```

File: docs/clients.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+On startup, client entry point scripts must first load the genesis block and state into
```
Comment (renaynay) [inline docs/clients.md:62]:
```suggestion
On startup, client entry-point scripts must first load the genesis block and state into
```

File: docs/clients.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+After the genesis state, the client should import the blocks from `/chain.rlp` if it is
+present, and finally import the individual blocks from `/blocks` in file name order. The
+reason for requiring two different block sources is that specifying a single chain is more
+optimal, but tests requiring forking chains cannot create a single chain. The client
+should start even if the blocks are invalid, i.e. after the import, the client's 'best
+block' should be the last valid, imported block.
+
+### Environment
+
+Clients must support the following environment variables. The client's entry point script
```
Comment (renaynay) [inline docs/clients.md:77]:
```suggestion
Clients must support the following environment variables. The client's entry-point script
```

File: docs/clients.md [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+| `HIVE_FORK_PETERSBURG`     | decimal              | [Petersburg][EIP-1716] transition block        |
+| `HIVE_FORK_ISTANBUL`       | decimal              | [Istanbul][EIP-1679] transition block          |
+| `HIVE_FORK_MUIRGLACIER`    | decimal              | [Muir Glacier][EIP-2387] transition block      |
+| `HIVE_FORK_BERLIN`         | decimal              | [Berlin][EIP-2070] transition block            |
+
+### Enode script
+
+Some tests require peer-to-peer node information of the client instance. The client
+container must contain an `/enode.sh` script that echoes the enode of the running
+instance. This script is executed by Hive host in order to retrieve the enode URL.
```
Comment (renaynay) [inline docs/clients.md:108]:
```suggestion
instance. This script is executed by the Hive host in order to retrieve the enode URL.
```

File: docs/commandline.md [hunk 8]
```diff
@@ -0,0 +1,101 @@
+[Overview] | [Hive Commands] | [Simulators] | [Clients]
+
+## Running Hive
+
+The hive project is implemented in Go. You need to install Go version 1.13 or later to use
+hive. To run simulations, you also need a working Docker setup, and hive needs to be run on the
+same machine as dockerd. Using docker remotely is not supported at this time. We have also
+not tested hive extensively on any OS but Linux, so you must run Linux to use hive.
+
+To get hive, you first need to clone to the repository to any location, then build the hive
```
Comment (renaynay) [inline docs/commandline.md:10]:
```suggestion
To get hive, you first need to clone the repository to any location, then build the hive
```

File: docs/commandline.md [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+The client list may contain any number of clients. You can select a specific client
+version by appending it to the client name with `_`, for example:
+
+    ./hive --sim devp2p/discv4 --client go-ethereum_v1.9.22,go-ethereum_v1.9.23
+
+Simulation runs can be customized in many ways. Here's an overview of the available
+command-line options.
+
+`--client.checktimelimit <timeout>`: The timeout of waiting for clients to open up TCP
+port 8545. If a very long chain is imported, this timeout may need to be quite large. A
```
Comment (renaynay) [inline docs/commandline.md:36]:
```suggestion
port 8545. If a very long chain is imported, this timeout may need to be quite long. A
```

File: docs/overview.md [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+## What is Hive?
+
+Hive is a system for running integration tests against Ethereum clients.
+
+In hive, integration tests are called 'simulations'. A simulation is controlled by a
+program (the 'simulator') written in any language. The simulator launches clients and
+contains test logic. It reports test results back to hive, where they are aggregated for
+display in a web browser.
+
+What makes hive different from other, generic CI infrastructure is the tight integration
```
Comment (renaynay) [inline docs/overview.md:12]:
```suggestion
What makes hive different from other generic CI infrastructure is the tight integration
```

File: docs/overview.md [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+
+![hive simulation docker containers](./img/hive-simulation.svg)
+
+Once all images are built, the simulator program is launched in a docker container. The
+`HIVE_SIMULATOR` environment variable contains the HTTP server URL of the hive controller.
+The [hive simulation API] can be accessed at this URL. The simulator launches clients and
+reports test results through the API.
+
+When the simulator requests a client instance, the hive controller launches a new docker
+container using the built client image. The client container entry point receives
```
Comment (renaynay) [inline docs/overview.md:83]:
```suggestion
container using the built client image. The client container entry-point receives
```

**Review by renaynay — APPROVED** [review]
Minor typos, but otherwise LGTM. It's very descriptive without going overboard.

File: docs/overview.md (hunk 10 above)
Comment (fjl) [reply docs/overview.md:12]: I want to keep this comma.

**Pull request merged**